src/bootstrap/check.rs

```diff
@@ -163,10 +163,43 @@ impl Step for Rustc {
     /// the `compiler` targeting the `target` architecture. The artifacts
     /// created will also be linked into the sysroot directory.
     fn run(self, builder: &Builder<'_>) {
-        let compiler = builder.compiler(0, builder.config.build);
+        let compiler = builder.compiler(builder.top_stage, builder.config.build);
         let target = self.target;
 
-        builder.ensure(Std { target });
+        if compiler.stage != 0 {
+            // If we're not in stage 0, then we won't have a std from the beta
+            // compiler around. That means we need to make sure there's one in
+            // the sysroot for the compiler to find. Otherwise, we're going to
+            // fail when building crates that need to generate code (e.g., build
+            // scripts and their dependencies).
+            builder.ensure(crate::compile::Std { target, compiler });
+
+            // Unfortunately, if rustc is not specifically requested, then we've
+            // already checked std too. That means that the sysroot has rmeta
+            // files lying around, which leads to duplicate crates being found
+            // by rustc.
+            //
+            // Ideally we'd somehow detect that we're going to end up building
+            // rustc (or some other tool) and stop the original check of std in
+            // stage 1, since it's not actually needed (we're building std just
+            // above), but that's hard, so instead we delete the artifacts from
+            // the sysroot here.
+            let libdir = builder.sysroot_libdir(compiler, target);
+            let hostdir = builder.sysroot_libdir(compiler, compiler.host);
+            let self_contained_dst = &libdir.join("self-contained");
+            for (path, dependency_type) in
+                builder.read_stamp_file(&libstd_stamp(builder, compiler, target))
+            {
+                let dst = match dependency_type {
+                    crate::DependencyType::Host => &*hostdir,
+                    crate::DependencyType::Target => &*libdir,
+                    crate::DependencyType::TargetSelfContained => &*self_contained_dst,
+                };
+                let _ = std::fs::remove_file(dst.join(path.file_name().unwrap()));
+            }
```

Comment:
Rather than adding the artifacts to sysroot just to delete them later, can you check std without adding it to sysroot? I think you'd only have to add `if builder.top_stage == 0` to https://github.com/rust-lang/rust/blob/937b96767095afdfc8341ecbf54bde1b965981e9/src/bootstrap/check.rs#L99.

I'm not too worried about the compile time, it just seems error-prone to delete only the sysroot files that are from libstd.